core/src/main/java/tachyon/worker/hierarchy/StorageDir.java

```diff
@@ -105,63 +115,125 @@
    * @throws IOException
    */
   public boolean cacheBlock(long userId, long blockId) throws IOException {
     String srcPath = getUserTempFilePath(userId, blockId);
     String dstPath = getBlockFilePath(blockId);
+    Long allocatedBytes = mTempBlockAllocatedBytes.remove(new Pair<Long, Long>(userId, blockId));
+
+    if (!mFs.exists(srcPath) || allocatedBytes == null) {
+      cancelBlock(userId, blockId);
+      throw new IOException("Block file doesn't exist! blockId:" + blockId);
+    }
     long blockSize = mFs.getFileSize(srcPath);
-    boolean result = mFs.rename(srcPath, dstPath);
-    if (result) {
-      addBlockId(blockId, blockSize);
+    if (blockSize < 0) {
+      cancelBlock(userId, blockId);
+      throw new IOException("Negative size of block! blockId:" + blockId);
+    }
+    returnSpace(userId, allocatedBytes - blockSize);
+    if (mFs.rename(srcPath, dstPath)) {
+      addBlock(blockId, blockSize, false);
+      updateUserOwnBytes(userId, -blockSize);
+      return true;
+    } else {
+      return false;
     }
-    return result;
   }
 
   /**
-   * Check status of the users, removedUsers can't be modified any more after being passed down from
-   * the caller
+   * Cancel a block which is being written
    * 
-   * @param removedUsers list of the removed users
+   * @param userId the id of the user
+   * @param blockId the id of the block to be cancelled
+   * @return true if success, false otherwise
+   * @throws IOException
    */
-  public void checkStatus(List<Long> removedUsers) {
-    for (long userId : removedUsers) {
-      Collection<Long> blockIds = mLockedBlocksPerUser.removeAll(userId);
-      for (long blockId : blockIds) {
-        mUserPerLockedBlock.remove(blockId, userId);
-      }
-      mUserAllocatedBytes.remove(userId);
+  public boolean cancelBlock(long userId, long blockId) throws IOException {  
+    String filePath = getUserTempFilePath(userId, blockId);
+    Long allocatedBytes = mTempBlockAllocatedBytes.remove(new Pair<Long, Long>(userId, blockId));
+    if (allocatedBytes == null) {
+      allocatedBytes = 0L;
+    }
+    returnSpace(userId, allocatedBytes);
+    if (!mFs.exists(filePath)) {
+      return true;
+    } else {
+      return mFs.delete(filePath, false);
     }
   }
 
   /**
+   * Clean resources related to the removed user
+   * 
+   * @param userId id of the removed user
+   * @param tempBlockIdList list of block ids that are being written by the user
+   */
+  public void cleanUserResources(long userId, Collection<Long> tempBlockIdList) {
+    Collection<Long> blockIds = mLockedBlocksPerUser.removeAll(userId);
+    for (long blockId : blockIds) {
+      mUserPerLockedBlock.remove(blockId, userId);
+    }
+    for (Long tempBlockId : tempBlockIdList) {
+      mTempBlockAllocatedBytes.remove(new Pair<Long, Long>(userId, tempBlockId));
+    }
+    try {
+      mFs.delete(getUserTempPath(userId), true);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```

Comment:
yes,  maybe logging error message instead of throwing exception is a better approach.